include/clad/Differentiator/ErrorEstimator.h

```diff
@@ -0,0 +1,89 @@
+#ifndef CLAD_ERROR_ESTIMATOR_H
+#define CLAD_ERROR_ESTIMATOR_H
+
+#include "EstimationModel.h"
+
+namespace clang {
+  class Stmt;
+  class Expr;
+  class Decl;
+  class VarDecl;
+  class FunctionDecl;
+  class CompoundStmt;
+} // namespace clang
+
+namespace clad {
+  struct DiffRequest;
+  class VisitorBase;
+  class DerivativeBuilder;
+  class FPErrorEstimationModel;
+} // namespace clad
+
+namespace clad {
+  /// The estimation handler which interfaces with Clad's reverse mode visitors
+  /// to fetch derivatives.
+  class ErrorEstimationHandler {
+    /// Keeps a track of the delta error expression we shouldn't emit.
+    bool m_DoNotEmitDelta = false;
+    /// Reference to the final error parameter in the augumented target
+    /// function.
+    clang::Expr* m_FinalError;
+    /// Reference to the return error expression.
+    clang::Expr* m_RetErrorExpr = nullptr;
+    /// A reference to the builder instance so that we can call Derive.
+    DerivativeBuilder& m_builder;
```

Comment:
```suggestion
    DerivativeBuilder& m_Builder;
```

Having the `ErrorEstimationHandler` drive `DerivativeBuilder` sounds like a design problem. To reduce the risk of turning this class into the God object antipattern we may be able to limit it by taking a function pointer to `Derive` and store it instead.

Reply:
No, I think you have misunderstood that. We only have a DerivativeBuilder reference there because we do not inherit from it (where we should, because our handler is a sort of visitor itself). We only use that instance call **```ReverseModeVisitor's```** derive. Again, I am only trying to "emulate" inheritance from ```DerivativeBuilder``` here, I am not taking over any control from it whatsoever. (Again, I am trying to achieve what ```JacobianModeVisitor``` already does).  For more info I think you can check out ```DerivativeBuilder's``` Derive and then ```ErrorEstimationHandler's``` ```Calculate```.

Reply:
For what it is worth, we can very well remove ```m_Builder``` from there and remove the ```Calculate``` function, just calling reverse mode Derive in the if-branch of ```DerivativeBuilder's``` Derive.

Reply:
Yes, I think that was what we agreed upon in the voice chat.